### PR TITLE
Fix building latest version of flex

### LINKED ISSUE #2398: Flex installation issues
I believe #2384 introduced some bugs to the flex installation. It now downloads from GitHub and runs autoreconf, but it must be missing some dependencies. On Fedora 24:
```
Can't exec "autopoint": No such file or directory at /home/ajstewart/spack/opt/spack/linux-fedora24-x86_64/gcc-6.2.1/autoconf-2.69-x5rbc5wjqydekkkjihgtjkoi7tzy6sco/share/autoconf/Autom4te/FileUtils.pm line 345.
autoreconf: failed to run autopoint: No such file or directory
autoreconf: autopoint is needed because this package uses Gettext
```
on CentOS 6:
```
autopoint: *** The AM_GNU_GETTEXT_VERSION declaration in your configure.ac file requires the infrastructure from gettext-0.19 but this version is older. Please upgrade to gettext-0.19 or newer.
autopoint: *** Stop.
autoreconf: autopoint failed with exit status: 1
```

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #2398. Also adds missing build dependencies for Autotools builds. The only problem is that I am unable to install `flex@2.5.39`, which is needed by FOAM-extend. For some reason it crashes because it can't find flex, but flex can't depend on itself, right?

### CODE REVIEW

**Comment by davydden** [comment]
@nrichart i think you contributed `foam-extend`, any hints on what could be going wrong here?

**Comment by adamjstewart** [comment]
Sorry, just realized my wording was confusing. I can't install `flex@2.5.39` because it can't find _itself_. Normally, I wouldn't care and would just remove older versions of flex, but FOAM-extend depends on an older version of flex, so I can't. FOAM-extend doesn't have any problems finding flex.

**Comment by adamjstewart** [comment]
But if @nrichart or anyone else could confirm that FOAM-extend can build with a newer version of flex, then all of my problems would go away.

File: var/spack/repos/builtin/packages/flex/package.py [hunk 1]
```diff
@@ -29,25 +29,41 @@ class Flex(AutotoolsPackage):
...
+    @when('@:2.6.0')
+    def autoreconf(self, spec, prefix):
+        libtoolize('--install', '--force')
+        autoreconf('--install', '--force')
+
+    @when('@2.6.1:')
     def autoreconf(self, spec, prefix):
-        autogen = Executable('./autogen.sh')
-        autogen()
+        pass
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/flex/package.py]: This is a known bug, dating back to #1490. You can't use `@when()` unless you specify it for every version.

Comment (alalazo) [inline var/spack/repos/builtin/packages/flex/package.py]: I don't think this is true. Try with this in the package:
```python
def autoreconf(self, spec, prefix):
    pass

@when('@:2.6.0')
def autoreconf(self, spec, prefix):
    libtoolize('--install', '--force')
￼    autoreconf('--install', '--force')
```
The first decorated method should be the default one, i.e. the one used if none of the decorated matches.

**Comment by davydden** [comment]
> I can't install flex@2.5.39 because it can't find itself. 

huh?! I am confused.

**Comment by adamjstewart** [comment]
> huh?! I am confused.

As am I...

The flex installation crashes during make:
```
/tmp/ajstewart/spack-stage/spack-stage-56QVui/flex-flex-2.5.39/missing: line 81: flex: command not found
WARNING: 'flex' is missing on your system.
         You should only need it if you modified a '.l' file.
         You may want to install the Fast Lexical Analyzer package:
         <http://flex.sourceforge.net/>
```

File: var/spack/repos/builtin/packages/flex/package.py (hunk 1 above)
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/flex/package.py]: It worked! That's interesting.

**Comment by adamjstewart** [comment]
Before my fix, I was unable to install Autoconf and Automake on Fedora 24. After my fix, I am unable to install Autconf and Automake on CentOS 6. God help me.

**Comment by adamjstewart** [comment]
Pretty sure the shebang is too long?
```
/bin/sh: ../bin/autom4te: /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1: bad interpreter: Permission denied
```
```
#! /blues/gpfs/home/software/spack-0.10.0/opt/spack/linux-centos6-x86_64/gcc-6.1.0/perl-5.24.0-4nllkpq46w752yd6mvw55q4lcv7tnkqu/bin/perl -w
```
Isn't there something that is supposed to shorten this?

**Comment by alalazo** [comment]
> Isn't there something that is supposed to shorten this?

@adamjstewart yes, it's in `lib/spack/spack/hooks/sbang.py`

**Comment by adamjstewart** [comment]
I think the shebang is generated during the build, so there's not much we can do about it. I'm just going to remove the perl dependencies for now. We've survived this long without them.

**Comment by adamjstewart** [comment]
Ok, removing the Perl dependency lets me build on CentOS 6, but prevents me from building on Fedora 24.

**Comment by nrichart** [comment]
For foam-extend we made it compile with flex >= 2.6. I just did not had time yet to test it on the upstream package.

**Comment by hegner** [comment]
Cross checked in our setup. Builds successfully on Scientific Linux 6 (=CentOS6) and CentOS 7. Thanks!

**Comment by adamjstewart** [comment]
Should I remove the version constraints from FOAM-extend and drop the older tarballs of flex that don't come with a configure script?

**Comment by adamjstewart** [comment]
Fixes #2427

**Comment by tgamblin** [comment]
@adamjstewart: Is this one ready to go?

**Comment by adamjstewart** [comment]
@tgamblin Probably. There's still the question of linking to Perl (which causes things to break on CentOS 6, but allows things to build on Fedora 24). There's also the question of whether or not we should just remove the old tarballs of Flex that require Autotools. But it builds for me now, and will help a lot of other users as is.

**Comment by tgamblin** [comment]
@adamjstewart: for the perl stuff, I'm curious why the file isn't getting patched by the `sbang` hook.  It walks the `prefix` after installation and looks for `#!` scripts, so there shouldn't be a problem if the file is generated during the build.  Is the issue that the file is generated *and run* during the build?

**Comment by adamjstewart** [comment]
>  Is the issue that the file is generated and run during the build?

Precisely.

**Comment by tgamblin** [comment]
@adamjstewart: seems like a more general problem than this PR -- and might need a more complicated fix.  I'll merge this for now.

Can you point me to the particular file and when it gets generated so I can look at this in the meantime?

**Comment by tgamblin** [comment]
@adamjstewart: Looks like putting "fixes #2401" in the comment does't actually close the issue on merge.  I think you have to edit the PR and add it at the top.  Also I've found that you can't do `Fixes #1, #2`.  You have to do `Fixes #1. Fixes #2.`

**Comment by adamjstewart** [comment]
Well, I just tested things again and I can actually build flex and friends on Fedora 24 without the Perl dependency. I still can't build the old tarball versions, but I won't need them. I think this is good for now.

**Comment by jrood-nrel** [comment]
I think I'm lost. Is this still being worked on? (I'm just trying to build binutils ultimately) I could easily be missing something, but I still cannot build flex with what seems like a fundamental error:

==> No patches needed for flex
==> Building flex [Package]
==> Executing phase : 'install'
==> Error: AttributeError: 'Flex' object has no attribute 'install'

**Comment by adamjstewart** [comment]
@jrood-nrel You definitely have something wrong there. The package type should be AutotoolsPackage, not Package. Did you apply these changes manually and forget a line? Are you on the latest develop branch?

**Comment by jrood-nrel** [comment]
Ah yes, I am bad at merging the development branch:

-class Flex(Package):
+class Flex(AutotoolsPackage):

Sorry about that!

**Pull request merged**